mcsema/Arch/ABI.cpp

```diff
@@ -540,60 +638,85 @@ void CallingConvention::StoreReturnValue(llvm::BasicBlock *block,
   }
 
   llvm::IRBuilder<> ir(block);
-  auto val_var = ReturnValVar(cc, val_type);
 
-  // If it's a pointer then convert it to a pointer-sized integer.
-  if (val_type->isPointerTy()) {
-    ret_val = ir.CreatePtrToInt(ret_val, gWordType);
-    val_type = gWordType;
-  }
+  for (unsigned i = 0; i < GetNumberOfElements(val_type); ++i) {
+    auto ret_type = RetrieveArgumentType(val_type, i);
```

Comment:
`RetrieveArgumentType` and `RetrieveArgumentType` can only really be used for `i` being `0` and `1`. There's no checking anywhere for the case where `i` is two or more, and if that happens, then funny things will happen.

I think if you make functions that take/return vectors (e.g. using `__attribute__((mode(XX))`, see [here](https://stackoverflow.com/a/4559049/247591)) then those would reasonably be passed in SSE registers (e.g. a vector of 16 `uint8_t`s), but this would ask for all of those values to be unpacked.

Reply:
In practice yeah, `i` will be probably only `0` or `1`, although if it become bigger, it will still be legal, since it only becomes bigger if `GetNumberOfElements` is bigger and therefore call to `RetrieveArgumentType` will still be legal. Or did I misunderstood the point?

Reply:
The points is `regs[index]` in `ReturnValVar`.

Here's an example: https://godbolt.org/g/N8u5xo

We have ```declare <4 x i32> @x() #1

declare { i64, i64 } @y() #1```

and `x`, when called, returns its value in `xmm0`, and `y` returns its values across `rax` and `rdx`.

Reply:
```asm
        call    x()
        movaps  XMMWORD PTR [rbp-16], xmm0
        call    y()
        mov     QWORD PTR [rbp-32], rax
        mov     QWORD PTR [rbp-24], rdx
```